File: src/main/java/conservative/ExchangeEngine.java
package conservative;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class ExchangeEngine {

    private final String currency;
    public ExchangeEngine(String currency){
        this.currency = currency;
    }

    public List<Models.Denomination> exchange(Integer amount) {
        assert(amount > 0);
        return recExchange(new ArrayList<>(), amount);
    }

    private List<Models.Denomination> recExchange(List<Models.Denomination> accumulator, Integer amount) {
        if (amount > 0) {
            Optional<Models.Denomination> nextDenom = largestDenominationFit(amount);
            Integer valueToSubtract = 0;
            if (nextDenom.isPresent()) {
                accumulator.add(nextDenom.get());
                valueToSubtract = nextDenom.get().getValue();
            }
            return recExchange(accumulator, amount - valueToSubtract);
        } else {
            return accumulator;
        }
    }

    private Optional<Models.Denomination> largestDenominationFit(Integer amount) {
        return DenominationSupplier.forCurrency(currency, DenominationSupplier.SortOrder.DESC)
                .stream().filter( dnm -> dnm.getValue() <= amount ).findFirst();
    }
}

File: src/main/kotlin/progressive/ExchangeEngine.kt
package progressive

class ExchangeEngine(private val currency: String) {
    fun exchange(amount: Int): List<Denomination> {
        assert(amount > 0)
        return recExchange(emptyList(), amount)
    }

    private fun recExchange(acc: List<Denomination>, amount: Int): List<Denomination> {
        return when (amount > 0) {
            true -> {
                val nextDenomination = largestDenominationFit(amount)
                recExchange(acc.plus(nextDenomination), amount - nextDenomination.value)
            }
            else -> acc
        }
    }

    private fun largestDenominationFit(amount: Int): Denomination {
        return DenominationSupplier.forCurrency(currency)
            .first { it.value <= amount }
    }
}
